### PR TITLE
add update timesheet flow

### PR DESCRIPTION
**Pull request by GGSukhorukovEgor**

### CODE REVIEW

File: src/endpoint/Timesheet.Update/Flow/FlowStep/Step.AwaitDate.cs [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+                        var activity = MessageFactory.Text(botFailure.UserMessage);
+                        await SendInsteadActivityAsync(context, cache.ActivityId, activity, cancellationToken);
+                        return context.RepeatSameStateJump<UpdateTimesheetFlowState>();
+                    })
+                .ToValueTask()
+                .ConfigureAwait(false);
+        }
+
+        var activity = MessageFactory.Text($"Выбери дату или введите дату в формате {DatePlaceholder}");
+        activity.ChannelData = JObject.FromObject(CreateChannelDataSelectDate(context.FlowState.Options.UrlWebApp));
```
Comment (Sauric) [inline src/endpoint/Timesheet.Update/Flow/FlowStep/Step.AwaitDate.cs]: Remove `JObject.FromObject`

File: src/endpoint/Timesheet.Update/Flow/FlowStep/Step.AwaitTimesheet.cs [hunk 2]
```diff
@@ -0,0 +1,430 @@
...
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace GarageGroup.Internal.Timesheet;
+partial class TimesheetUpdateFlowStep
```
Comment (Sauric) [inline src/endpoint/Timesheet.Update/Flow/FlowStep/Step.AwaitTimesheet.cs:13]: Add empty line

File: src/endpoint/Timesheet.Update/Flow/FlowStep/Step.AwaitTimesheet.cs [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+
+        return ChatFlowJump.Repeat<UpdateTimesheetFlowState>(
+            new DateWebAppCacheJson(resource.Id, UpdateStatus.ProjectIsEdited, default, project, cache.EditTimesheetState));
+    }
+
+    private static async ValueTask<ChatFlowJump<UpdateTimesheetFlowState>> SuccessSelectTimesheetAsync(
+        (IChatFlowContext<UpdateTimesheetFlowState> Context, UpdateTimesheetJson Timesheet, ICrmProjectApi CrmProjectApi, DateWebAppCacheJson Cache) input,
+        CancellationToken cancellationToken = default)
+    {
+        return (input.Timesheet, input.Cache) switch
```
Comment (Sauric) [inline src/endpoint/Timesheet.Update/Flow/FlowStep/Step.AwaitTimesheet.cs]: Use inline return `=>`

File: src/endpoint/Timesheet.Update/Flow/FlowStep/Step.AwaitTimesheet.cs [hunk 4]
```diff
@@ -0,0 +1,430 @@
...
+
+    private static async ValueTask<ChatFlowJump<UpdateTimesheetFlowState>> SelectTimesheet(
+        this ICrmProjectApi crmProjectApi,
+        IChatFlowContext<UpdateTimesheetFlowState> context, 
+        CancellationToken token)
+    {
+        var turnContext = (ITurnContext)context;
+        var cache = context.StepState as DateWebAppCacheJson;
+
+        if (cache != null)
```
Comment (Sauric) [inline src/endpoint/Timesheet.Update/Flow/FlowStep/Step.AwaitTimesheet.cs]: Use pattern match `if (context.StepState is DateWebAppCacheJson cache)`

File: src/endpoint/Timesheet.Update/Flow/FlowStep/Step.AwaitTimesheet.cs [hunk 5]
```diff
@@ -0,0 +1,430 @@
...
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace GarageGroup.Internal.Timesheet;
+partial class TimesheetUpdateFlowStep
+{
+    internal static ChatFlow<UpdateTimesheetFlowState> AwaitTimesheetWebApp(this ChatFlow<UpdateTimesheetFlowState> chatFlow, ICrmProjectApi crmProjectApi)
+        =>
+        chatFlow.ForwardValue(crmProjectApi.SelectTimesheet);
```
Comment (Sauric) [inline src/endpoint/Timesheet.Update/Flow/FlowStep/Step.AwaitTimesheet.cs]: Move to next line
``` csharp
=>
chatFlow.ForwardValue(
    crmProjectApi.SelectTimesheet);
```

File: src/endpoint/Timesheet.Update/Flow/FlowStep/Step.AwaitTimesheet.cs [hunk 6]
```diff
@@ -0,0 +1,430 @@
...
+                    UpdateProject = true
+                },
+                _ => ChatFlowJump.Break<UpdateTimesheetFlowState>(ChatFlowBreakState.From(UnknownErrorText))
+            };
+        }
+
+        return await SelectDate(context, turnContext, token).ConfigureAwait(false);
+    }
+
+    private static async Task<ChatFlowJump<UpdateTimesheetFlowState>> ProcessingMessagesWebApp(
```
Comment (Sauric) [inline src/endpoint/Timesheet.Update/Flow/FlowStep/Step.AwaitTimesheet.cs]: Use `ValueTask` type and remove `async`

File: src/endpoint/Timesheet.Update/Flow/FlowStep/Step.AwaitTimesheet.cs [hunk 7]
```diff
@@ -0,0 +1,430 @@
...
+        ICrmProjectApi crmProjectApi, 
+        IChatFlowContext<UpdateTimesheetFlowState> context, 
+        DateWebAppCacheJson cache, 
+        CancellationToken token)
+        =>
+        await AsyncPipeline
```
Comment (Sauric) [inline src/endpoint/Timesheet.Update/Flow/FlowStep/Step.AwaitTimesheet.cs]: Remove `await`

File: src/endpoint/Timesheet.Update/Flow/FlowStep/Step.AwaitTimesheet.cs [hunk 8]
```diff
@@ -0,0 +1,430 @@
...
+                data => (context, data, crmProjectApi, cache))
+            .FoldValue(
+                SuccessSelectTimesheetAsync,
+                async (botFailure, cancellationToken) =>
+                {
+                    var activity = MessageFactory.Text(botFailure.UserMessage);
+                    await SendInsteadActivityAsync(context, cache.ActivityId, activity, cancellationToken).ConfigureAwait(false);
+                    return context.RepeatSameStateJump<UpdateTimesheetFlowState>();
+                })
+            .ToValueTask();
```
Comment (Sauric) [inline src/endpoint/Timesheet.Update/Flow/FlowStep/Step.AwaitTimesheet.cs]: Remove `.ToValueTask()`

File: src/endpoint/Timesheet.Delete/FlowStep/Step.AwaitDate.cs [hunk 9]
```diff
@@ -38,15 +38,17 @@ internal static ChatFlow<DeleteTimesheetFlowState> AwaitDateWebApp(this ChatFlow
...
                     })
                 .ToValueTask()
                 .ConfigureAwait(false);
         }
 
         var activity = MessageFactory.Text($"Выбери дату или введите дату в формате {DatePlaceholder}");
-        activity.ChannelData = JObject.FromObject(CreateChannelDataSelectDate(context.FlowState.Options.UrlWebApp));
+
+        var daysInterval = context.FlowState.Options.TimesheetInterval.Days;
+        activity.ChannelData = JObject.FromObject(CreateChannelDataSelectDate(context.FlowState.Options.UrlWebApp, daysInterval));
```
Comment (Sauric) [inline src/endpoint/Timesheet.Delete/FlowStep/Step.AwaitDate.cs]: Remove `JObject.FromObject`

File: src/endpoint/Timesheet.Delete/FlowStep/Step.AwaitTimesheet.cs [hunk 10]
```diff
@@ -34,10 +34,11 @@ internal static ChatFlow<DeleteTimesheetFlowState> AwaitTimesheetWebApp(this Cha
                     async (botFailure, cancellationToken) =>
                     {
                         var activity = MessageFactory.Text(botFailure.UserMessage);
-                        await SendInsteadActivityAsync(context, cache.ActivityId, activity, cancellationToken);
+                        await SendInsteadActivityAsync(context, cache.ActivityId, activity, cancellationToken).ConfigureAwait(false);
                         return context.RepeatSameStateJump<DeleteTimesheetFlowState>();
                     })
-                .ToTask();
+                .ToTask()
```
Comment (Sauric) [inline src/endpoint/Timesheet.Delete/FlowStep/Step.AwaitTimesheet.cs]: Use `ToValueTask` instead

File: src/endpoint/Timesheet.Update/Flow/FlowState/UpdateTimesheetFlowState.cs [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+    [JsonProperty("timesheets")]
+    public IReadOnlyCollection<TimesheetJson>? Timesheets { get; init; }
+
+    public UpdateTimesheetJson? TimesheetUpdate { get; init; }
+
+    public bool UpdateProject { get; init; }
+
+    public UpdateTimesheetOptions Options { get; init; }
+
+    public UpdateTimesheetFlowState(UpdateTimesheetOptions options)
```
Comment (pmosk) [inline src/endpoint/Timesheet.Update/Flow/FlowState/UpdateTimesheetFlowState.cs:54]: Remove this constructor

File: src/endpoint/Timesheet.Update/Flow/FlowStep/Step.ReadContextData.cs [hunk 12]
```diff
@@ -0,0 +1,47 @@
+using GarageGroup.Infra.Bot.Builder;
+using Microsoft.Bot.Builder;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace GarageGroup.Internal.Timesheet;
+
+partial class TimesheetUpdateFlowStep
```
Comment (pmosk) [inline src/endpoint/Timesheet.Update/Flow/FlowStep/Step.ReadContextData.cs]: Delete this step

**Review by pmosk — CHANGES_REQUESTED** [review]

File: src/endpoint/Timesheet.Update/Flow/FlowStep/Step.UpdateTimesheet.cs [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+                   ) : null,
+            duration: editedTimesheet.Duration != timesheet.Duration ? editedTimesheet.Duration : null,
+            description: editedTimesheet.Description != timesheet.Description ? editedTimesheet.Description : null,
+            timesheetId: editedTimesheet.Id);
+    }
+
+    private static ChatFlowBreakState ToBreakState(Failure<TimesheetUpdateFailureCode> failure)
+         =>
+         ("Ну удалось изменить запись.")
+         .Pipe(
```
Comment (pmosk) [inline src/endpoint/Timesheet.Update/Flow/FlowStep/Step.UpdateTimesheet.cs]: Don't use pipeline here

File: src/endpoint/Timesheet.Update/Flow/Internal.Json/ProjectCacheJson.cs [hunk 14]
```diff
@@ -0,0 +1,13 @@
...
+namespace GarageGroup.Internal.Timesheet;
+
+internal sealed record ProjectCacheJson
+{
+    public Guid Id { get; init; }
+
+    public string? Name { get; init; }
+
+    public string? Data { get; init; }
+
```
Comment (pmosk) [inline src/endpoint/Timesheet.Update/Flow/Internal.Json/ProjectCacheJson.cs]: remove empty line

**Review by pmosk — CHANGES_REQUESTED** [review]

**Review by pmosk — APPROVED** [review]

**Pull request merged**